### PR TITLE
Add option to check if plugin try to set viewer attr outside main thread

### PR DESCRIPTION
**Pull request by Czaki**
# Description

When reading #5142 I remind that in the past, we met problems that may lead to application crashes if the viewer attribute is modified outside the main thread. 
In this PR I add `NAPARI_ENSURE_PLUGIN_MAIN_THREAD` environment variable that, if set, enables the check of the thread from which the object is modified. This modification works only for objects wrapped with `PublicOnlyProxy`, which is how the viewer is passed to plugins in napari. Not all modifications outside the main thread may be caught. It is possible, that use of some napari function may lead to this (as the wrapper is removed when moving to the napari code).

This PR introduces dependency of `napari.utils._proxies` from `napari._qt` but only if the environment variable is set. 

Add this PR to the main will allow the plugin creator to check if they do such an operation. For example, when they observe a random crash of napari. Example of such crash could be found here: #4982

This PR also warns to set private attributes in `PublicOnlyProxy` wrapped object. Currently, we only warn about getting its value. 

This PR requires to add documentation in `docs/plugins/debug_plugins.md] from #5142. So it may require an update if #5142 is merged first. 

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Review by jni — APPROVED** [review]

File: napari/_qt/utils.py
Comment (andy-sweet) [inline napari/_qt/utils.py]: Nit: when I see `check` I think of functions that will `raise` if a condition isn't met. Instead this returns a boolean, in which case something shorter is probably more clear and less verbose.

```suggestion
def in_main_thread() -> bool:
```

File: napari/utils/_proxies.py
Comment (andy-sweet) [inline napari/utils/_proxies.py]: Is this a leftover?

File: napari/utils/_proxies.py
Comment (andy-sweet) [inline napari/utils/_proxies.py]: It's fairly unclear what this is checking. In `__getattr__` we use the same logic, but it's clarified with an explanatory/implementation comment. Ideally we should refactor this into a function with a meaningful name (`called_from_napari`), though we'll have to be careful about the `depth` passed to `_getframe`.

File: napari/utils/_tests/test_proxies.py
Comment (andy-sweet) [inline napari/utils/_tests/test_proxies.py]: Nice improvements.

File: napari/utils/_tests/test_proxies.py
Comment (andy-sweet) [inline napari/utils/_tests/test_proxies.py]: Nit: What is the `ex` short for? Maybe `exception` or `example`? Would be good to expand it to clarify.

File: napari/_qt/utils.py
Comment (andy-sweet) [inline napari/_qt/utils.py:485]: Could we rely on Python's `threading.current_thread` and `threading.main_thread` to implement this function in a Qt-independent way and define it outside of `_qt`?

File: napari/utils/_proxies.py
Comment (andy-sweet) [inline napari/utils/_proxies.py]: Given this class is named `PublicOnlyProxy`, I'm not sure this is the right place for this check. But it is convenient and I can't think of a better approach right now. I also can't think of a good name that would describe both checks, so I think it's fine to leave it as is for now - especially since it's guarded with the environment variable.

File: napari/utils/_tests/test_proxies.py
Comment (andy-sweet) [inline napari/utils/_tests/test_proxies.py]: I'd recommend implementing this with a `ThreadPoolExecutor` fixture instead of using threads directly because the test implementation will be shorter and more readable (especially with respect to the exception). Something like the following.

```python
@pytest.mark.skipif("qtpy" not in sys.modules, reason="requires Qt")
def test_thread_proxy_guard(monkeypatch, qapp, single_threaded_executor):
    class X:
        a = 1

    monkeypatch.setenv('NAPARI_ENSURE_PLUGIN_MAIN_THREAD', 'True')

    x = X()
    x_proxy = PublicOnlyProxy(x)

    f = single_threaded_executor.submit(x.__setattr__, 'a', 2)
    f.result()
    assert x.a == 2

    f = single_threaded_executor.submit(x_proxy.__setattr__, 'a', 3)
    with pytest.raises(RuntimeError):
        f.result()
    assert x.a == 2
```

**Review by andy-sweet — CHANGES_REQUESTED** [review]
This looks good, though I'd like to see if we could do this without Qt and there may be an opportunity simplify a few things in the implementation and tests. None of my comments are strong, so feel free to push back and I'll likely approve.

Ideally, this would be two PRs - one for the private `__setattr__` warning and one for the non-main thread `__setattr__` exception. I guess the non-main thread code is only executed if an environment variable is defined, so I'd be OK bringing this altogether.

File: napari/_qt/utils.py
Comment (Czaki) [reply napari/_qt/utils.py:485]: I do not think earlier about main_thread. 

In general it is not good solution (as main thread in python is thread in which start interpreter, and Qt main threat is thread where Application object is created. 

But as it is for debugging purpose, then, with proper explanation it could be done in this way. I need to rethink this.

File: napari/utils/_proxies.py
Comment (Czaki) [reply napari/utils/_proxies.py]: same as in `__getattr__`. When we remove private access there should be raised an exception about this,

File: napari/utils/_tests/test_proxies.py
Comment (Czaki) [reply napari/utils/_tests/test_proxies.py]: exception, I will change

File: napari/utils/_tests/test_proxies.py
Comment (Czaki) [reply napari/utils/_tests/test_proxies.py]: Where this fixture is implemented?

File: napari/_qt/utils.py
Comment (andy-sweet) [reply napari/_qt/utils.py:485]: If we expect that the interpreter's main thread to sometimes be different from the thread in which the application is created, then maybe this should be called `in_qt_main_thread` or `in_gui_thread`?

Comment (Czaki) [reply napari/_qt/utils.py:485]: I do not expect but this is possible. But I'm not sure if we need to support it.

File: napari/utils/_proxies.py
Comment (Czaki) [reply napari/utils/_proxies.py]: done

File: napari/utils/_tests/test_proxies.py
Comment (Czaki) [reply napari/utils/_tests/test_proxies.py]: done

File: napari/_qt/utils.py
Comment (perlman) [reply napari/_qt/utils.py:485]: I second Andy's sentiments here. Mostly, it'd be nice to new Qt dependancies in otherwise non-graphical code.

File: napari/utils/_proxies.py
Comment (perlman) [inline napari/utils/_proxies.py]: Is this something that could go into settings instead?
(e.g. how experimental octtree is configured)

I'm not a fan of having behavior deep into napari changed based on external configurations.  (But, I don't have a strong conviction here. :)

Comment (Czaki) [reply napari/utils/_proxies.py]: > Is this something that could go into settings instead?

But this is for debugging purposes for the plugin creators. Not something that a common user may want to use.

File: napari/_qt/utils.py
Comment (Czaki) [reply napari/_qt/utils.py:485]: So you suggest implementing this in pure python? So, where put the documentation that is required to create a QApplication object in the python main thread?

File: napari/utils/_proxies.py
Comment (perlman) [reply napari/utils/_proxies.py]: Perhaps a dev section in the settings? The target -- devs or users -- doesn't change my feeling one continuous querying of the environment.

File: napari/_qt/utils.py
Comment (perlman) [reply napari/_qt/utils.py:485]: I don't know.  I don't have strong feelings about it, especially with this sitting in a `qt` directory.

File: napari/utils/_proxies.py
Comment (Czaki) [reply napari/utils/_proxies.py]: Checking documentation https://docs.python.org/3/library/os.html#os.environ, and it is a pure dict query. It will be faster than querying settings.

**Comment by Czaki** [comment]
I have implemented an alternative approach when Qt check will be used only if available, but the code will work even if Qt is unavailable.

File: napari/_qt/_tests/test_qt_utils.py
Comment (andy-sweet) [inline napari/_qt/_tests/test_qt_utils.py]: Probably best to make this a fixture that calls `shutdown` on teardown (credit to @kcpevey): 

```python
@pytest.fixture()
def single_threaded_executor():
    executor = ThreadPoolExecutor(max_workers=1)
    yield executor
    executor.shutdown()
```

Or use the executor in a context when it's needed:

```python
with ThreadPoolExecutor() as executor:
    f = executor.submit(x.__setattr__, 'a', 2)
    f.result()
    assert x.a == 2

    f = executor.submit(x_proxy.__setattr__, 'a', 3)
    with pytest.raises(RuntimeError):
        f.result()
```

File: napari/utils/_proxies.py
Comment (andy-sweet) [inline napari/utils/_proxies.py]: I love how readable this is now!

File: napari/utils/_proxies.py
Comment (andy-sweet) [inline napari/utils/_proxies.py:169]: Why do we need the global? Is this so that we only execute this function once for performance reasons?

This function is only called when an env variable is set, so I don't think we need to worry to much about performance. So if that's the only reason to do this, I'd suggest removing the global and just defining this function as `in_main_thread`.

**Comment by andy-sweet** [comment]
The warning for `setattr` looks great and should definitely come into v0.4.17. How strongly do you feel about bringing in the thread checks? I still feel like splitting this PR in two could be the best way to go here.

File: napari/utils/_proxies.py
Comment (Czaki) [reply napari/utils/_proxies.py:169]: Yes, this is for performance reasons.

Comment (andy-sweet) [reply napari/utils/_proxies.py:169]: Then I'd definitely suggest simplify this to:

```python
def in_main_thread() -> bool:
    try:
        from napari._qt.utils import in_qt_main_thread

        return in_qt_main_thread()
    except ImportError:
        return in_main_thread_py()
    except AttributeError:
        warnings.warn(
            "Qt libs are available but no QtApplication instance is created"
        )
        return in_main_thread_py()
```

Does that make sense?

Comment (Czaki) [reply napari/utils/_proxies.py:169]: Yes

**Comment by jni** [comment]
@Czaki Sorry, could you remove the changes to the `docs/` folder and make a separate PR over at https://github.com/napari/docs?

**Comment by Czaki** [comment]
> @Czaki Sorry, could you remove the changes to the `docs/` folder and make a separate PR over at https://github.com/napari/docs?

I will do this

File: napari/utils/_testsupport.py
Comment (andy-sweet) [inline napari/utils/_testsupport.py:265]: Is this used anywhere? If not, did you use it to help debug this PR?

If it's not used, but you think it's useful, you might want to add a comment to clarify that this is useful for debugging and should not be deleted.

**Review by andy-sweet — APPROVED** [review]
Looks good - thanks for the updates!

File: napari/utils/_testsupport.py
Comment (Czaki) [reply napari/utils/_testsupport.py:265]: this is a replacement for `make_napari_viewer` for plugins creators and is referred in my PR to docs https://github.com/napari/docs/pull/7/files 

Currently, plugins during tests do not have warnings about private attribute access, only during actual execution.

**Comment by Czaki** [comment]
merge this requires https://github.com/napari/docs/pull/7 to be merged also

**Comment by andy-sweet** [comment]
@Czaki : as this has two PRs to coordinate (this + docs), feel free to merge both when you'r ready.

**Comment by Czaki** [comment]
> @Czaki : as this has two PRs to coordinate (this + docs), feel free to merge both when you'r ready.

@andy-sweet I have added a test. Could you check it?

File: napari/utils/_tests/test_proxies.py
Comment (andy-sweet) [inline napari/utils/_tests/test_proxies.py]: Optional: As I read it, this looks like `test_thread_proxy_guard` but with the napari viewer. If that's the case, I'd consider matching the style of those with executors rather than using `Thread`:

```suggestion
def test_proxy_fixture_thread_error(make_napari_viewer_proxy, single_threaded_executor):
    viewer = make_napari_viewer_proxy()
    future = single_threaded_executor.submit(viewer.__setattr__, 'status', 'hi')
    with pytest.raises(RuntimeError, match='Setting attributes'):
        future.result()
```

For me, that makes it clearer what this is testing as the `with pytest.raises` statement doesn't get so buried. Whereas in the original form `assert ok` seems like the main thing we're testing, whereas that's mostly checking that `with pytest.raises` swallows the right exception.

I guess the downsides of using the executor are that it doesn't actually raise the exception until you call `result`  and the ugliness of `__setattr__`. I'll leave it up to you.

Comment (andy-sweet) [reply napari/utils/_tests/test_proxies.py]: Whatever you choose, fix the name of the test though - it currently ends in `threat_warning`, whereas I think it should be something like `thread_error`.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
New test looks good - except for the name of it! I'll approve after you change that. 

I also suggested an alternate more concise implementation of the test. That's optional.

File: napari/utils/_tests/test_proxies.py
Comment (Czaki) [reply napari/utils/_tests/test_proxies.py]: Your implementation is shorter so I prefer it.

**Review by andy-sweet — APPROVED** [review]
Looks great! Feel free to merge when you want.

File: napari/plugins/_tests/test_npe2.py
Comment (lucyleeow) [inline napari/plugins/_tests/test_npe2.py:24]: @Czaki could you please explain why this is needed (I can't work it out)? Thank you!

**Pull request merged**
